### PR TITLE
[DOC] Update api stability documentation.

### PR DESCRIPTION
**Pull request by smehringer**
A first try. Please review.

Resolves https://github.com/seqan/product_backlog/issues/279

### CODE REVIEW

File: doc/about/api/index.md [hunk 1]
```diff
@@ -19,11 +19,12 @@ Exceptions to the previous rule:
      `detail` or if it is private to a class type, you are not allowed to depend upon it any way. More generally,
```
Comment (eseiler) [inline doc/about/api/index.md:19]: In line 1: `future-proofness`. Apparently, that's not really a word. There is `future-proof` and `future-proofing`, but not `future-proofness`.

File: doc/about/api/index.md [hunk 2]
```diff
@@ -19,11 +19,12 @@ Exceptions to the previous rule:
...
      if it is not part of the *user documentation*, it is not part of the API.
   2. **Do not depend on entities marked as "experimental" or "NOAPI"**. Major changes to the library like new modules
-     are often marked as experimental within the first minor release they appear in. This means we do not guarantee
-     stability until the next minor release happens and the experimental flag is removed. Very few entities in namespace
-     `seqan3` are permanently marked as "NOAPI" which designates them as subject to unannounced change. This is usually
-     the case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate
-     correct API documentation for entities in `seqan3`.
+     are often first marked as experimental within the a minor release cycle. This means we do not guarantee
```
Comment (eseiler) [inline doc/about/api/index.md:22]:
```suggestion
     are often first marked as experimental within a minor release cycle. This means we do not guarantee
```

File: doc/about/api/index.md [hunk 3]
```diff
@@ -19,11 +19,12 @@ Exceptions to the previous rule:
...
+     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.
+     Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to
+     unannounced change. In addition all entities in **core** and **utlity** are always marked "NOAPI". This is the
```
Comment (eseiler) [inline doc/about/api/index.md:25]:
```suggestion
     unannounced change. In addition, all entities in **core** and **utility** are always marked "NOAPI". This is the
```

File: doc/about/api/index.md [hunk 4]
```diff
@@ -19,11 +19,12 @@ Exceptions to the previous rule:
...
+     case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate correct
+     API documentation for entities in `seqan3` or considered nice-to-have general features for users.
   4. **Do not depend on the *signatures* of SeqAn APIs.** In particular you may not take the address of APIs in SeqAn
```
Comment (eseiler) [inline doc/about/api/index.md:28]:
```suggestion
  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular, you may not take the address of APIs in SeqAn
```

File: doc/about/api/index.md [hunk 5]
```diff
@@ -74,11 +75,15 @@ a different (updated) on-disk format.
 # Platform stability {#platform_stability}
 
 The main requirement for being able to use SeqAn is a supported compiler and in general all operating systems that
-provide a supported compiler should be supported. We currently support the
-following compilers on 64-bit operating systems with little-endian CPU
-architectures only:
+provide a supported compiler should be supported. In general, we only support the lates three major compiler versions.
```
Comment (eseiler) [inline doc/about/api/index.md:78]:
```suggestion
provide a supported compiler should be supported. In general, we only support the latest three major compiler versions.
```

File: doc/about/api/index.md [hunk 6]
```diff
@@ -74,11 +75,15 @@ a different (updated) on-disk format.
...
+
+We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures only:
```
Comment (eseiler) [inline doc/about/api/index.md:80]:
```suggestion
We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures:
```

File: doc/about/api/index.md [hunk 7]
```diff
@@ -74,11 +75,15 @@ a different (updated) on-disk format.
...
   * GCC7, GCC8, GCC9, GCC10, GCC11
 
+\note Only the most recent minor release of a compiler is guaranteed to be supported, e.g. when `gcc-10.4` is released,
+we may drop support for `gcc-10.3`. Since all platforms with an older version receive minor release updates,
+this should not be a problem.
+
 We promise to support all of the above compilers until the last release of SeqAn3, or until all of the following
```
Comment (eseiler) [inline doc/about/api/index.md:87]:
```suggestion
We promise to support all the above compilers until the last release of SeqAn3, or until all the following
```

**Review by eseiler — CHANGES_REQUESTED** [review]
First look

File: doc/about/api/index.md (hunk 1 above)
Comment (smehringer) [reply doc/about/api/index.md:19]: I came up with 

1. `# API Stability - future-proof promises of seqan3`
2. `# Future-proof API Stability`
3. `# Stability and long term promises`

Thoughts?

File: doc/about/api/index.md [hunk 8]
```diff
@@ -19,12 +19,13 @@ Exceptions to the previous rule:
      `detail` or if it is private to a class type, you are not allowed to depend upon it any way. More generally,
```
Comment (SGSSGene) [inline doc/about/api/index.md:19]: Nicht sicher, mein "Sprachgefühl" sagt mir, dort muss noch ein "in" oder ähnliches rein.
```suggestion
     `detail` or if it is private to a class type, you are not allowed to depend upon it in any way. More generally,
```

File: doc/about/api/index.md [hunk 9]
```diff
@@ -74,12 +75,16 @@ a different (updated) on-disk format.
...
+
+We currently support the following compilers on 64-bit operating systems with little-endian CPU architectures:
   * GCC7, GCC8, GCC9, GCC10, GCC11
 
-We promise to support all of the above compilers until the last release of SeqAn3, or until all of the following
+\note Only the most recent minor release of a compiler is guaranteed to be supported, e.g. when `gcc-10.4` is released,
+we may drop support for `gcc-10.3`. Since all platforms with an older version receive minor release updates,
+this should not be a problem.
+
+We promise to support all the above compilers until the last release of SeqAn3, or until all the following
```
Comment (SGSSGene) [inline doc/about/api/index.md:87]: Nicht ganz sicher was damit ausgedrückt werden soll, vielleicht so?:
```suggestion
We promise to support the above compilers in the latest release of SeqAn3, or until all the following
```

**Review by SGSSGene — CHANGES_REQUESTED** [review]

File: doc/about/api/index.md [hunk 10]
```diff
@@ -19,12 +19,13 @@ Exceptions to the previous rule:
...
-     `seqan3` are permanently marked as "NOAPI" which designates them as subject to unannounced change. This is usually
-     the case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate
-     correct API documentation for entities in `seqan3`.
-  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular you may not take the address of APIs in SeqAn
+     are often first marked as experimental within a minor release cycle. This means we do not guarantee
+     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.
+     Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to
+     unannounced change. In addition, all entities in **core** and **utility** are always marked "NOAPI". This is the
+     case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate correct
+     API documentation for entities in `seqan3` or considered nice-to-have general features for users.
```
[the remark this answers is not in the document]
Comment (smehringer) [reply doc/about/api/index.md:27]: agree but I changed it a little differently

File: doc/about/api/index.md (hunk 1 above)
Comment (eseiler) [reply doc/about/api/index.md]: I like 3, though it should be `long-term`

Comment (eseiler) [reply doc/about/api/index.md]: > Ich finde 1. am besten. Begründung: 2. ist kaum unterscheidbar zu "# API stability" 3. weiß nicht, 1. gefällt mir besser :p

But `future-proof promises` means that the promises are future-proof, no?

File: doc/about/api/index.md (hunk 9 above)
Comment (eseiler) [reply doc/about/api/index.md:87]: I think it wants to express `Until SeqAn 4.0.0`. Though that's probably not correct anyway :)

File: doc/about/api/index.md [hunk 11]
```diff
@@ -19,12 +19,13 @@ Exceptions to the previous rule:
...
+  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular, you may not take the address of APIs in SeqAn
```
[the remark this answers is not in the document]
Comment (smehringer) [reply doc/about/api/index.md:28]: changed everythin to automated numbering

File: doc/about/api/index.md (hunk 1 above)
Comment (smehringer) [reply doc/about/api/index.md]: I kind of like 3 the most too. But I don't care much.

Comment (SGSSGene) [reply doc/about/api/index.md]: 2:1 for option three, I guess it is settled :-)

File: doc/about/api/index.md [hunk 12]
```diff
@@ -54,9 +55,9 @@ update path:
   * The next (or another future) minor version removes the old API.
 
 \warning
-As a special case of point 2. **the entire 3.0.* release is marked as experimental**.
-We will do our best not to break things, but similar to the releases of GCC, you should consider `seqan-3.1.0` as
-the first API-stable release of SeqAn3.
+As a special case of point 2. **the entire 3.0.* release is not stable**.
```
Comment (eseiler) [inline doc/about/api/index.md:58]:
```suggestion
As a special case of point 2, **the entire 3.0.* release is not stable**.
```

**Review by eseiler — CHANGES_REQUESTED** [review]
💅

File: doc/about/api/index.md [hunk 13]
```diff
@@ -15,32 +15,33 @@ release cycle of SeqAn3, i.e. if you write an application that includes `seqan-3
 to re-compile it against `seqan-3.4.5` without errors.
```
[the remark this answers is not in the document]
Comment (smehringer) [reply doc/about/api/index.md:15]: I think you are only right about non-stable stuff. I think entities marked as stable that become deprecated at some point we promise to leave them in the code base until a new major release.

[the remark this answers is not in the document]
Comment (eseiler) [reply doc/about/api/index.md]: Yes, that's also right, stable stuff will stay until 4.0.0
But experimental entities might be deprecated and later removed?

[the remark this answers is not in the document]
Comment (smehringer) [reply doc/about/api/index.md]: yes but that is already covered in the notes that follow ("do not depend on experimental stuff")

File: doc/about/api/index.md [hunk 14]
```diff
@@ -1,46 +1,48 @@
...
-     are often marked as experimental within the first minor release they appear in. This means we do not guarantee
-     stability until the next minor release happens and the experimental flag is removed. Very few entities in namespace
-     `seqan3` are permanently marked as "NOAPI" which designates them as subject to unannounced change. This is usually
-     the case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate
-     correct API documentation for entities in `seqan3`.
-  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular you may not take the address of APIs in SeqAn
+  -# **Do not depend on entities marked as "experimental" or "no-api"**. Major changes to the library like new modules
+     are often first marked as experimental within a minor release cycle. This means we do not guarantee
+     stability until the next minor release happens and the experimental flag is removed in favour of the stable flag.
+     Some entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to
```
Comment (eseiler) [inline doc/about/api/index.md:25]:
```suggestion
     Some entities in namespace `seqan3` are permanently marked as "no-api" which designates them as subject to
```

**Review by eseiler — APPROVED** [review]

**Pull request merged**